### ISSUE TITLE
Tunnel URL not found error in fresh expo init

### ISSUE THREAD

**Issue by FernandoKGA** [expo/expo-cli#2459]
## Description
The `--tunnel` option isn't working properly, it cannot find the Tunnel URL, it always fallback to LAN URL, throwing the warning `Tunnel URL not found (it might not be ready yet), falling back to LAN URL.`.

## Expected Behavior
I have performed the following command: `expo start --tunnel --dev` inside the blank project I had just created, and I expected to start Expo and Metro Bundler with the Tunnel option working correctly.

## Observed Behavior
When starting Expo, it already prompts the error of Tunnel URL not found, then waiting more 2 minutes, nothing happens and more warnings are prompted, then changing to LAN and returning to Tunnel, it prompts `Error starting tunnel` and fails again.

When scanning the QR Code, it works normally, as it is in LAN, using the Expo App in a Xiaomi Redmi Note 7.

```
user@user-desktop:~/app$ expo start --tunnel --dev
Starting project at /home/user/app
Expo DevTools is running at http://localhost:19002
Opening DevTools in the browser... (press shift-d to disable)
Starting Metro Bundler.
Tunnel URL not found (it might not be ready yet), falling back to LAN URL.
Tunnel URL not found (it might not be ready yet), falling back to LAN URL.
Tunnel URL not found (it might not be ready yet), falling back to LAN URL.
Tunnel URL not found (it might not be ready yet), falling back to LAN URL.

  exp://192.168.0.39:19000

...QR Code here...

To run the app with live reloading, choose one of:
  • Sign in as @_XXX_ in Expo client on Android or iOS. Your projects will automatically appear in the "Projects" tab.
  • Scan the QR code above with the Expo app (Android) or the Camera app (iOS).
  • Press a for Android emulator, or w to run on web.
  • Press e to send a link to your phone with email.

 Expo  Press ? to show a list of all available commands.
Logs for your project will appear below. Press Ctrl+C to exit.
Tunnel URL not found (it might not be ready yet), falling back to LAN URL.
Tunnel URL not found (it might not be ready yet), falling back to LAN URL.
Tunnel URL not found (it might not be ready yet), falling back to LAN URL.
Tunnel URL not found (it might not be ready yet), falling back to LAN URL.
Tunnel URL not found (it might not be ready yet), falling back to LAN URL.
Error starting tunnel 
Tunnel URL not found (it might not be ready yet), falling back to LAN URL.
Tunnel URL not found (it might not be ready yet), falling back to LAN URL.
```

## Environment
```
Expo CLI 3.24.0 environment info:
    System:
      OS: Linux 5.4 Ubuntu 20.04.1 LTS (Focal Fossa)
      Shell: 5.0.17 - /bin/bash
    Binaries:
      Node: 12.18.3 - ~/.nvm/versions/node/v12.18.3/bin/node
      Yarn: 1.22.4 - /usr/bin/yarn
      npm: 6.14.6 - ~/.nvm/versions/node/v12.18.3/bin/npm
    IDEs:
      Android Studio: 4.0 AI-193.6911.18.40.6626763
    npmPackages:
      expo: ~38.0.8 => 38.0.9 
      react: ~16.11.0 => 16.11.0 
      react-dom: ~16.11.0 => 16.11.0 
      react-native: https://github.com/expo/react-native/archive/sdk-38.0.2.tar.gz => 0.62.2 
      react-native-web: ~0.11.7 => 0.11.7 
    npmGlobalPackages:
      expo-cli: 3.24.0
```
Phone/emulator/simulator platform, model and version:
- Xiaomi Redmi Note 7
- Android 10 QKQ1.190910.002
- MIUI Global 11.0.1.0(QFGMIXM)

## Reproducible Demo
Install a fresh `expo-cli`, run `expo init app`, `cd app` to enter the folder and run `expo start --tunnel --dev`.

**Comment by Immortalin** [comment]
Logout and login to your expo again (from expo-cli)

**Comment by pierrefrota** [comment]
@FernandoKGA  I have the same problem, managed to solve?

**Comment by FernandoKGA** [comment]
> @FernandoKGA I have the same problem, managed to solve?

Not yet, I will test tonight again and I will give some feedback about it. Have been focusing in other tasks at my job...

**Comment by pierrefrota** [comment]
@FernandoKGA did you test?

**Comment by Immortalin** [comment]
Try logout of expo CLI and login again

**Comment by pierrefrota** [comment]
@Immortalin  I already did that, it didn't work. when I click on the 'tunnel' button, it displays the following message in devTools in the browser: ```Error loading DevTools
ValidationError: "urlType" must be one of [exp, http, redirect, no-protocol]```

**Comment by baba43** [comment]
@pierrefrota I've the same issue I guess.

https://stackoverflow.com/questions/64802001/expo-tunnel-not-working-invalid-certificate

**Comment by pierrefrota** [comment]
@baba43 my packager-info.json:
`{
  "devToolsPort": 19002,
  "expoServerPort": 19000,
  "packagerPort": 19001,
  "packagerPid": 15202,
  "expoServerNgrokUrl": null,
  "packagerNgrokUrl": null,
  "ngrokPid": null
}
`

**Comment by pierrefrota** [comment]
I did the test on a Windows PC. The tunnel worked perfectly. The problem is in some configuration of my Ubuntu.

**Comment by pierrefrota** [comment]
a new message on my terminal:

`Tunnel URL not found (it might not be ready yet), falling back to LAN URL.
XDLError: ValidationError: "urlType" must be one of [exp, http, redirect, no-protocol]
at assertValidOptions (/@expo/xdl@58.0.19/src/UrlUtils.ts:218:11)
at ensureOptionsAsync (/@expo/xdl@58.0.19/src/UrlUtils.ts:228:5)
at constructUrlAsync (/@expo/xdl@58.0.19/src/UrlUtils.ts:280:22)
at constructUrlAsync (/@expo/xdl@58.0.19/src/UrlUtils.ts:334:14)
at Object.constructHostUriAsync (/@expo/xdl@58.0.19/src/UrlUtils.ts:47:21)
at getManifestResponseAsync (/@expo/xdl@58.0.19/src/project/ManifestHandler.ts:218:22)
at /@expo/xdl@58.0.19/src/project/ManifestHandler.ts:130:49`

**Comment by otaviogaiao** [comment]
Same issue here on Mac OS catalina

**Comment by pierrefrota** [comment]
```sudo yarn start``` it worked for me

**Comment by JustinLawrenceMS** [comment]
sudo yarn start worked for me on MacOS.  This is not good.  We should fix this.

**Comment by jjborie** [comment]
Same issue here on windows.

**Comment by FernandoKGA** [comment]
Ok... So... I have analyzed the code where the error is logged, at `packages/xdl/src/UrlUtils.ts`. That part of the code is from the method `constructUrlAsync` at L279, it uses the variable `isPackager` that is provided to the method to check if the `proxyURL` comes from either `process.env.EXPO_PACKAGER_PROXY_URL` or `process.env.EXPO_MANIFEST_PROXY_URL`, and underneath that, is checked if the variable `proxyURL` exists. In this case, the variable don't exist, so the code will analyze the type of host it was supposed to be, yours is `'tunnel'`, leading to: https://github.com/expo/expo-cli/blob/964c47928559cad2b30f386770dc11dc87c17048/packages/xdl/src/UrlUtils.ts#L329

The URLs the `packagerInfo` variable provides in it's properties fails the check, leading to the log warning.

**Comment by FernandoKGA** [comment]
> Ok... So... I have analyzed the code where the error is logged, at `packages/xdl/src/UrlUtils.ts`. That part of the code is from the method `constructUrlAsync` at L279, it uses the variable `isPackager` that is provided to the method to check if the `proxyURL` comes from either `process.env.EXPO_PACKAGER_PROXY_URL` or `process.env.EXPO_MANIFEST_PROXY_URL`, and underneath that, is checked if the variable `proxyURL` exists. In this case, the variable don't exist, so the code will analyze the type of host it was supposed to be, yours is `'tunnel'`, leading to:
> 
> https://github.com/expo/expo-cli/blob/964c47928559cad2b30f386770dc11dc87c17048/packages/xdl/src/UrlUtils.ts#L329
> 
> The URLs the `packagerInfo` variable provides in it's properties fails the check, leading to the log warning.

Going to debug to seek for more information.

**Comment by FernandoKGA** [comment]
So.. I had updated the version to the newest expo-cli version in order to test it, as the problem keep happening.

```log
Expo CLI 4.0.6 environment info:
    System:
      OS: Linux 5.4 Ubuntu 20.04.1 LTS (Focal Fossa)
      Shell: 5.0.17 - /bin/bash
    Binaries:
      Node: 12.18.3 - ~/.nvm/versions/node/v12.18.3/bin/node
      Yarn: 1.22.5 - ~/.yarn/bin/yarn
      npm: 6.14.6 - ~/.nvm/versions/node/v12.18.3/bin/npm
    IDEs:
      Android Studio: 4.0 AI-193.6911.18.40.6626763
    npmPackages:
      expo: ~39.0.2 => 39.0.5 
      react: 16.13.1 => 16.13.1 
      react-dom: 16.13.1 => 16.13.1 
      react-native: https://github.com/expo/react-native/archive/sdk-39.0.4.tar.gz => 0.63.2 
      react-native-web: ~0.13.12 => 0.13.18 
    npmGlobalPackages:
      expo-cli: 4.0.6
    Expo Workflow: managed
```

In this method: https://github.com/expo/expo-cli/blob/964c47928559cad2b30f386770dc11dc87c17048/packages/xdl/src/UrlUtils.ts#L204
```js
const schema = joi.object().keys({
    scheme: joi.string().optional().allow(null),
    // Replaced by `scheme`
    urlType: joi.any().valid('exp', 'http', 'redirect', 'no-protocol'),
```

```log
ferkarchiloff@ferkarchiloff-desktop:~/my-app$ yarn start --tunnel
yarn run v1.22.5
$ expo start --tunnel
Starting project at /home/ferkarchiloff/my-app
Expo DevTools is running at http://localhost:19002
Opening DevTools in the browser... (press shift-d to disable)
projectRoot /home/ferkarchiloff/my-app
incomingOpts { urlType: 'http', hostType: 'localhost' }
isPackager true
requestHostname undefined
Starting Metro Bundler
projectRoot /home/ferkarchiloff/my-app
incomingOpts null
isPackager false
requestHostname undefined
Tunnel URL not found (it might not be ready yet), falling back to LAN URL.
projectRoot /home/ferkarchiloff/my-app
incomingOpts {
  urlType: null,
  scheme: null,
  hostType: 'lan',
  lanType: 'ip',
  dev: true,
  minify: false,
  urlRandomness: null,
  https: false
}
isPackager false
requestHostname undefined
ValidationError: "urlType" must be one of [exp, http, redirect, no-protocol]
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

The schema is validating the urlType, the first time it enters it, it passes correctly. The second run it doesn't find the `incomingOpts`, so it will not run in the schema, then it will run the code in:
https://github.com/expo/expo-cli/blob/964c47928559cad2b30f386770dc11dc87c17048/packages/xdl/src/UrlUtils.ts#L235

After this run, it's probably where the problem happens, because the object gotten is: 
```js
incomingOpts {
  urlType: null,
  scheme: null,
  hostType: 'lan',
  lanType: 'ip',
  dev: true,
  minify: false,
  urlRandomness: null,
  https: false
}
```

However, the input in `settings.json` I've configured is: 
```js
{
  "urlType": "exp",
  "scheme": null,
  "hostType": "tunnel",
  "lanType": "ip",
  "dev": true,
  "minify": false,
  "urlRandomness": null,
  "https": false
}
```
This object is not being read. But probably the configuration returned by default is causing the program to fail. The original problem related to the tunnel URL not being ready, remains.

**Comment by brentvatne** [comment]
delete the `.expo` directory in your project if you get a validation error. related: https://github.com/expo/expo-cli/issues/2952

as for the tunnel not starting, try installing [ngrok](https://ngrok.com/) and testing it with an http server (like `npx serve` for example). that may help with debugging, because we use ngrok to start the tunnel.

**Comment by FernandoKGA** [comment]
> delete the `.expo` directory in your project if you get a validation error. related: #2952
> 
> as for the tunnel not starting, try installing [ngrok](https://ngrok.com/) and testing it with an http server (like `npx serve` for example). that may help with debugging, because we use ngrok to start the tunnel.

I had deleted the `.expo` folder and tried to initiate the project normally with `yarn start`, then changed to Tunnel option, and it failed showing the `urlType` error message. As a second test, I'd deleted it again and used `--tunnel` option, which failed again.

The `ngrok` installed in my computer via `snap` is working fine, tested it with `npx serve`, working normally as it should be.

**Comment by brentvatne** [comment]
maybe we should provide a way to let the user provide a path to their own ngrok binary... do you want to open a PR to add support for that?

**Comment by FernandoKGA** [comment]
> maybe we should provide a way to let the user provide a path to their own ngrok binary... do you want to open a PR to add support for that?

I can try, probably will take some time because I need to study the code and imagine how to provide this way, but I can try.

**Comment by prince4799** [comment]
I had the same issue but then I stop the metro bundler and again  run the project  it works fine after several tries.

**Comment by brunoarueira** [comment]
If the path for ngrok binary will be the way to fix this issue, probably is better to sync https://github.com/expo/ngrok from it's upstream, because https://github.com/bubenshchykov/ngrok already support a custom bin path. See:

https://github.com/bubenshchykov/ngrok/blob/b2cc8d0503d7d2495b86fb30885dc8845a601d88/process.js#L36

PS: I've the same issue

**Comment by marinoandrea** [comment]
Since version `4.1.3` I am no longer getting the `urlType` error that @pierrefrota was referring to. 
However, Expo is now falling back to LAN url whenever I try to start it with the `--tunnel` option and I also noticed a new warning displayed in DevTools whenever I switch to tunnel mode:

`Error starting tunnel The system cannot find the path specified.`

I am not sure what the warning is about, but here's my environment:

```    
Expo CLI 4.1.4 environment info:
    System:
      OS: Windows 10 10.0.19041
    Binaries:
      Node: 14.15.1 - C:\Program Files\nodejs\node.EXE
      Yarn: 1.22.10 - C:\Users\****\AppData\Roaming\npm\yarn.CMD
      npm: 6.14.8 - C:\Program Files\nodejs\npm.CMD
    IDEs:
      Android Studio: Version  4.1.0.0 AI-201.8743.12.41.6953283
    npmPackages:
      expo: ^40.0.0 => 40.0.0
      react: 16.13.1 => 16.13.1
      react-dom: 16.13.1 => 16.13.1
      react-native: https://github.com/expo/react-native/archive/sdk-40.0.1.tar.gz => 0.63.2 
      react-native-web: ~0.13.12 => 0.13.18
    Expo Workflow: managed
```  

On a side note, I have experienced the same behaviour on WSL Ubuntu 20.04.

**Comment by FernandoKGA** [comment]
I'm going to close this issue as it seems to be resolved in newer versions. I have installed `ngrok` through the package manager `snap` and it worked out of the box with a brand new project created through the CLI. Do you see any problems in closing this issue @brentvatne ?

Diagnostics info:
```bash
Expo CLI 5.0.5 environment info:
    System:
      OS: Linux 5.11 Ubuntu 20.04.3 LTS (Focal Fossa)
      Shell: 5.0.17 - /bin/bash
    Binaries:
      Node: 14.17.3 - ~/.nvm/versions/node/v14.17.3/bin/node
      Yarn: 1.22.17 - /usr/bin/yarn
      npm: 8.2.0 - ~/.nvm/versions/node/v14.17.3/bin/npm
    npmPackages:
      expo: ~44.0.0 => 44.0.4 
      react: 17.0.1 => 17.0.1 
      react-dom: 17.0.1 => 17.0.1 
      react-native: 0.64.3 => 0.64.3 
      react-native-web: 0.17.1 => 0.17.1 
    Expo Workflow: managed
```

**Comment by Jarrettgohh** [comment]
I was getting the same error `Tunnel URL not found (it might not be ready yet), falling back to LAN URL.`. 

1. I managed to fix it by first running `expo start` in one terminal. 
2. I got the LAN URL, and created a tunnel myself by running `ngrok http <LAN URL>` -> for example `ngrok http http://192.168.1.90:19000`. 
3.  Next, I visited the URL given by ngok and suffixed it with `/_expo/loading` -> example `https://c5e1-42-61-215-17.ap.ngrok.io/_expo/loading` and loaded the app from there

**Comment by Kodamace** [comment]
Hey Guys this is what I did to just get it running on my physical device. Emulator etc is different I don't know if this is really that helpful but I have tried a couple things but in the end the way I got it to run on my device is like this

1. npx expo start --tunnel
  After running this I then scanned the qr code with my physical device but this tries to open a browser at http:// which doesn't work 
2. I then just entered the url manually in my expo go app to connect exp:// url not the http:// url 

This is located in your terminal
Metro waiting on exp://url.exp.redirect

The only way I was able to get this running on my physical device
